### PR TITLE
Tiles chart component

### PR DESCRIPTION
**Pull request by inanda**
Created - Tiles chart Component

- The component scales the tiles based on the number of tiles and the chart width and height props
- Tile colors, maximum size, minimum size and data is configurable
- The component was added to the demo page as an example of use

See how it looks like.

tileschart

### CODE REVIEW

File: src/ws-tiles-chart/tile.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import {Component, PropTypes} from "../imports";
+
+
+
+export const Tile = ({data, config, size}) => {
```
Comment (fragsalat) [inline src/ws-tiles-chart/tile.js]: JSdocs missing

File: src/ws-tiles-chart/tile.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import {Component, PropTypes} from "../imports";
```
Comment (fragsalat) [inline src/ws-tiles-chart/tile.js]: Could you please also import React here? The jsx transpiles to React.createElement and if you don't add React to the window scope this won't work.

File: src/ws-tiles-chart/tile.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+export const Tile = ({data, config, size}) => {
+  const style = {
+    'backgroundColor': config.backgroundColor,
+    width: `${size}px`,
+    height: `${size}px`
+  };
+
+  return (
+    <div key={data.key} className="tile" style={style}/>
```
Comment (fragsalat) [inline src/ws-tiles-chart/tile.js]: I think you don't need to add the key attribute here

File: src/ws-tiles-chart/tile.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  );
+};
+
+Tile.propTypes = {
```
Comment (fragsalat) [inline src/ws-tiles-chart/tile.js]: Could you pls use the static keyword in the class instead

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
+import {Component, PropTypes} from "../imports";
+import {Tile} from "./tile";
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: We're using single quotes instead of double quotes :)

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * This class describes a Preact/React component which renders a tiles chart
+ * @class WSTilesChart
+ * @property {Object} props React properties object
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: The possible properties should be documented here. Pls check other components like ws-dropdown

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  /**
+   * @param {Object} props React properties
+   * @constructor
+   */
+  constructor(props) {
+    super(props);
+    this.state = {tileSize: 0};
+
+    this.calculateMaximumPossibleTileSize = this.calculateMaximumPossibleTileSize.bind(
+      this);
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: You don't need to wrap here. We have a right margin on 120 characters

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  componentDidMount() {
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: JSDoc missing :)

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    //Calculate the tiles size
+    const tilesAmount = this.props.data.groups.map(group=> {
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: White space before => missing

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      return group.tilesSet.length
+    }).reduce((a, b)=>a + b);
+
+    const tileSize = this.calculateMaximumPossibleTileSize(this.tilesChartContainer.clientWidth,
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: Could you pls wrap if the maximum chars per line are exceeded like this?
```js
func(
  param1,
  param2,
  param3
)
```

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      this.tilesChartContainer.clientHeight, tilesAmount);
+    this.setState({
+      tileSize: (tileSize<this.props.maxTileSize)? tileSize:this.props.maxTileSize
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: Whitespaces?^^

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+  /**
+   * Calculates the maximum tile size based on total width, height and amount of tiles
+   * @param {width} chart container width
+   * @param {height} chart container height
+   * @returns {int}
+   */
+  calculateMaximumPossibleTileSize(width, height, tilesAmount) {
+    const chartArea = (width < height) ? width * width : height * height;
+    return Math.sqrt((chartArea) / tilesAmount)
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: Unnecessary braces around charArea

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+   * Renders the chart
+   * @returns {Object}
+   */
+  render() {
+    const {data, config, title} = this.props;
+    return (
+      <div className="ws-tiles-chart">
+        <div className="tiles-chart-title">{title}</div>
+
+        <div className="tiles-chart-container"
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: If this isto long pls wrap like this
```html
<div
  className="tiles-chart-container"
  ref="asdasd"
>
  <span></span>
</div>
```

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+             ref={(element)=>this.tilesChartContainer=element}>
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: You really don't like whitespaces or?^^

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+          {data.groups.map(group=>group.tilesSet.map(
+            tile => {
+              const groupConfig = config.groupConfigs
+                ? config.groupConfigs.find(config=>config.key === group.key)
+                : {};
+              return <Tile data={tile} config={groupConfig}
+                           size={this.state.tileSize}/>
+            }
+          ))}
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: could be written like that
```js
{data.groups.map(group => group.titlesSet.map(tile => {
  <Tile
    data={tile}
    config={([] || config.groupConfigs.find(config => config.key === group.key) || {}}
    size={this.state.tileSize}
  />
}}
```

or even more clear

```js
const getGroupConfig = (config, group) => {
  if (config.groupConfigs) {
    return config.groupConfigs.find(config => config.key === group.key);
  }
  return {};
}
return (
  <div>
    {data.groups.map(group => group.titlesSet.map(tile => {
      <Tile
        data={tile}
        config={getGroupConfig(config, group)}
        size={this.state.tileSize}
      />
    }}
  </div>
)
```

File: src/ws-tiles-chart/ws-tiles-chart.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        </div>
+
```
Comment (fragsalat) [inline src/ws-tiles-chart/ws-tiles-chart.js]: Useless line :)

**Comment by inanda** [comment]
I did not know about the gulp lint, gulp build. Next time I will for sure run it and fix all the spaces, etc, before opening the pull request :D.

After the fixes, the circleCI tests passed.
I've also added some additional changes, like making minTileSize, width and height configurable by properties.

File: demo/demo_app.js [hunk 17]
```diff
@@ -76,6 +77,10 @@ render(
           </tr>
         </tbody>
       </table>
+      <br />
+        <WSTilesChart
+          title="Chart title" data={dashboardMockData.data} config={dashboardMockData.config} width={100} height={100}
```
Comment (fragsalat) [inline demo/demo_app.js]: This wrap is not needed or wrong wrapped^^

File: demo/mockdata.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
+export const dashboardMockData = {
+  data: {
+    groups: [{
+      key: 'lessThan2h',
+      tilesSet: [{key: '1'}, {key: '2'}, {key: '3'}, {key: '4'}, {key: '5'}]
+    },
+      {
```
Comment (fragsalat) [inline demo/mockdata.js]: Wrong indention

**Pull request merged**